src/main/java/io/github/hotlava03/chatutils/listeners/AntiSpamListener.java

```diff
@@ -1,34 +1,102 @@
 package io.github.hotlava03.chatutils.listeners;
 
+import io.github.hotlava03.chatutils.events.ReceiveMessageEvent;
 import io.github.hotlava03.chatutils.fileio.ChatStorage;
 import io.github.hotlava03.chatutils.fileio.ChatUtilsConfig;
-import io.github.hotlava03.chatutils.events.ReceiveMessageEvent;
-import io.github.hotlava03.chatutils.util.StringUtils;
-import net.minecraft.text.Text;
+import net.minecraft.client.MinecraftClient;
+import net.minecraft.client.util.ChatMessages;
+import net.minecraft.util.math.MathHelper;
 
 import java.util.function.Consumer;
 
-public class AntiSpamListener implements Consumer<ReceiveMessageEvent> {
-    private Text lastMessage;
-    private int spamCounter = 1;
+import static io.github.hotlava03.chatutils.util.OrderedTextAdapter.orderedTextToString;
+import static io.github.hotlava03.chatutils.util.StringUtils.getDifference;
+import static io.github.hotlava03.chatutils.util.StringUtils.isNumeric;
 
+public class AntiSpamListener implements Consumer<ReceiveMessageEvent> {
     @Override
     public void accept(ReceiveMessageEvent e) {
         if (ChatStorage.getInstance().isBlockingChatEvents()) return;
-        if (ChatUtilsConfig.ANTI_SPAM.value()) {
-            if (this.lastMessage == null) {
-                this.lastMessage = e.getText();
-                this.spamCounter = 1;
-                return;
+        var client = MinecraftClient.getInstance();
+        var chat = client.inGameHud.getChatHud();
+        var serverInfo = client.getCurrentServerEntry();
+        var address = serverInfo != null ? serverInfo.address : null;
+        var fullHistory = e.getLines();
+        var range = ChatUtilsConfig.ANTI_SPAM_RANGE.value();
+        var history = fullHistory.size() >= range ? fullHistory.subList(0, range) : fullHistory;
+        if (history.isEmpty()) return;
+
+        int maxTextLength = MathHelper.floor(chat.getWidth() / chat.getChatScale());
+        var splitLines = ChatMessages.breakRenderedChatMessageLines(
+                e.getText(), maxTextLength, client.textRenderer);
+
+        int spamCounter = 1;
+        int lineMatchCount = 0;
```

Comment:
3 `int`s in a sea of `var`s, nice consistency
```suggestion
        var maxTextLength = MathHelper.floor(chat.getWidth() / chat.getChatScale());
        var splitLines = ChatMessages.breakRenderedChatMessageLines(
                e.getText(), maxTextLength, client.textRenderer);

        var spamCounter = 1;
        var lineMatchCount = 0;
```